tsl/src/chunk.c

```diff
@@ -186,3 +206,889 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type, boo
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("merging compressed chunks is not yet supported"),
+					 errhint("Decompress the chunks before merging.")));
+
+		if (hypertable_id == INVALID_HYPERTABLE_ID)
+			hypertable_id = chunk->fd.hypertable_id;
+		else if (hypertable_id != chunk->fd.hypertable_id)
+		{
+			Assert(i > 0);
+
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("cannot merge chunks across different hypertables"),
+					 errdetail("Chunk \"%s\" is part of hypertable \"%s\" while chunk \"%s\" is "
+							   "part of hypertable \"%s\"",
+							   get_rel_name(chunk->table_id),
+							   get_rel_name(chunk->hypertable_relid),
+							   get_rel_name(relinfos[i - 1].chunk->table_id),
+							   get_rel_name(relinfos[i - 1].chunk->hypertable_relid))));
+		}
+
+		/* Need to check that storage is compatible across the chunks being
+		 * merged. It is not possible to merge two chunks with different
+		 * storage, except for Hypercore TAM and Heap TAM, which are
+		 * compatible (since both using heap storage). */
+		amoid = rel->rd_rel->relam;
+
+		if (amoid != HEAP_TABLE_AM_OID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unsupported access method for merge")));
+
+		relinfos[i].relid = relid;
+		relinfos[i].rel = rel;
+		relinfos[i].chunk = chunk;
+	}
+
+	/* Sort rels in partition order (in case of chunks). This is necessary to
+	 * validate that a merge is possible. */
+	qsort(relinfos, nrelids, sizeof(RelationMergeInfo), cmp_relations);
+
+	/* Step 2: Check alignment/mergeability and create the merged hypercube
+	 * (partition ranges). */
+	for (int i = 0; i < nrelids; i++)
+	{
+		const Chunk *chunk = relinfos[i].chunk;
+
+		Assert(chunk != NULL);
+
+		if (merged_cube == NULL)
+		{
+			merged_cube = ts_hypercube_copy(chunk->cube);
+			Assert(prev_cube == NULL);
+		}
+		else
+		{
+			Assert(chunk->cube->num_slices == merged_cube->num_slices);
+			Assert(prev_cube != NULL);
+			validate_merge_possible(prev_cube, chunk->cube);
+			merge_cubes(merged_cube, chunk->cube);
+		}
+
+		prev_cube = chunk->cube;
+		compute_rel_vacuum_cutoffs(relinfos[i].rel, &relinfos[i].cutoffs);
+	}
+
+	Relation result_rel = relinfos[0].rel;
+	/* These will be our final cutoffs for the merged relation */
+	struct VacuumCutoffs *cutoffs = &relinfos[0].cutoffs;
+
+	/* Keep the first relation of the ones merged. It will receive a heap swap. */
+	Oid tablespace = result_rel->rd_rel->reltablespace;
+	relpersistence = result_rel->rd_rel->relpersistence;
+
+	/* Create the transient heap that will receive the re-ordered data */
+#if PG15_LT
+	Oid new_relid =
+		make_new_heap(RelationGetRelid(result_rel), tablespace, relpersistence, ExclusiveLock);
+#else
+	Oid am = result_rel->rd_rel->relam;
+	Oid new_relid =
+		make_new_heap(RelationGetRelid(result_rel), tablespace, am, relpersistence, ExclusiveLock);
+#endif
```

Comment:
Shouldn't this `make_new_heap` be added to compat.h and create `make_new_heap_compat` macro?